File: solutions/go/rotational-cipher/1/rotational_cipher.go
package rotationalcipher

import "unicode"

func RotationalCipher(plain string, shiftKey int) (r string) {
	for _, c := range plain {
		switch {
		case unicode.IsLower(c):
			c = ((c - 'a' + rune(shiftKey)) % 26) + 'a'
		case unicode.IsUpper(c):
			c = ((c - 'A' + rune(shiftKey)) % 26) + 'A'
		}
		r += string(c)
	}
	return
}

File: solutions/haskell/rotational-cipher/1/src/RotationalCipher.hs
module RotationalCipher (rotate) where

import Data.Char (chr, isLower, isUpper, ord)

rotate :: Int -> String -> String
rotate key = map rot
  where
    rot c
        | isLower c = rot2 'z'
        | isUpper c = rot2 'Z'
        | otherwise = c
      where
        c1 = ord c + key
        rot2 z = chr $ if c1 > ord z then c1 - 26 else c1
